diff --git a/include/linux/memcontrol.h b/include/linux/memcontrol.h
index 20ff87f8e..97cbf3a35 100644
--- a/include/linux/memcontrol.h
+++ b/include/linux/memcontrol.h
@@ -210,6 +210,8 @@ struct mem_cgroup {
 		struct page_counter memsw;	/* v1 only */
 	};
 
+	struct page_counter anonmem;	/* v1 only */
+	
 	/* Legacy consumer-oriented counters */
 	struct page_counter kmem;		/* v1 only */
 	struct page_counter tcpmem;		/* v1 only */
diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index 61932c921..b9073ec03 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -210,6 +210,7 @@ enum res_type {
 	_MEMSWAP,
 	_KMEM,
 	_TCP,
+	_ANON,
 };
 
 #define MEMFILE_PRIVATE(x, val)	((x) << 16 | (val))
@@ -3906,6 +3907,9 @@ static u64 mem_cgroup_read_u64(struct cgroup_subsys_state *css,
 	case _TCP:
 		counter = &memcg->tcpmem;
 		break;
+	case _ANON:
+		counter = &memcg->anonmem;
+		break;
 	default:
 		BUG();
 	}
@@ -4075,6 +4079,9 @@ static ssize_t mem_cgroup_write(struct kernfs_open_file *of,
 		case _TCP:
 			ret = memcg_update_tcp_max(memcg, nr_pages);
 			break;
+		case _ANON:
+			ret = -EINVAL;
+			break;
 		}
 		break;
 	case RES_SOFT_LIMIT:
@@ -4108,6 +4115,9 @@ static ssize_t mem_cgroup_reset(struct kernfs_open_file *of, char *buf,
 	case _TCP:
 		counter = &memcg->tcpmem;
 		break;
+	case _ANON:
+		counter = &memcg->anonmem;
+		break;
 	default:
 		BUG();
 	}
@@ -5210,6 +5220,17 @@ static int mem_cgroup_slab_show(struct seq_file *m, void *p)
 static int memory_stat_show(struct seq_file *m, void *v);
 
 static struct cftype mem_cgroup_legacy_files[] = {
+    {
+        .name = "anonmem.usage_in_bytes",
+		.private = MEMFILE_PRIVATE(_ANON, RES_USAGE),
+		.read_u64 = mem_cgroup_read_u64,
+    },
+    {
+        .name = "anonmem.limit_in_bytes", 
+        .read_u64 = mem_cgroup_read_u64,
+        .write = mem_cgroup_write,
+        .private = MEMFILE_PRIVATE(_ANON, RES_LIMIT),
+    },
 	{
 		.name = "usage_in_bytes",
 		.private = MEMFILE_PRIVATE(_MEM, RES_USAGE),
@@ -5584,12 +5605,14 @@ mem_cgroup_css_alloc(struct cgroup_subsys_state *parent_css)
 		page_counter_init(&memcg->swap, &parent->swap);
 		page_counter_init(&memcg->kmem, &parent->kmem);
 		page_counter_init(&memcg->tcpmem, &parent->tcpmem);
+		page_counter_init(&memcg->anonmem, &parent->anonmem);
 	} else {
 		init_memcg_events();
 		page_counter_init(&memcg->memory, NULL);
 		page_counter_init(&memcg->swap, NULL);
 		page_counter_init(&memcg->kmem, NULL);
 		page_counter_init(&memcg->tcpmem, NULL);
+		page_counter_init(&memcg->anonmem, NULL);
 
 		root_mem_cgroup = memcg;
 		return &memcg->css;
@@ -5738,6 +5761,7 @@ static void mem_cgroup_css_reset(struct cgroup_subsys_state *css)
 	page_counter_set_max(&memcg->swap, PAGE_COUNTER_MAX);
 	page_counter_set_max(&memcg->kmem, PAGE_COUNTER_MAX);
 	page_counter_set_max(&memcg->tcpmem, PAGE_COUNTER_MAX);
+	page_counter_set_max(&memcg->anonmem, PAGE_COUNTER_MAX);
 	page_counter_set_min(&memcg->memory, 0);
 	page_counter_set_low(&memcg->memory, 0);
 	page_counter_set_high(&memcg->memory, PAGE_COUNTER_MAX);
